File: ruby/medium/longest_substring_without_repeating_characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    map = {}
    i, j, longest = 0, 0, 0
    while j < s.length
        c = s[j]
        if map.has_key?(c) && map[c] >= i
            i = map[c] + 1
        end
        longest = [longest, (j-i+1)].max
        map[c] = j
        j += 1
    end
    return longest
end

File: javascript/medium/longestSubstringWithoutRepeatingCharacters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let i = 0, j = 0, longest = 0;
    const map = new Map();
    while(j < s.length) {
        const c = s[j];
        if(map.has(c) && map.get(c) >= i) {
            i = (map.get(c) + 1);
        }
        longest = Math.max(longest, (j-i+1));
        map.set(c, j);
        j++;
    }
    return longest;
};
